source/common/router/upstream_request.cc

```diff
@@ -93,7 +94,8 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
       cleaned_up_(false), had_upstream_(false),
       stream_options_({can_send_early_data, can_use_http3}), grpc_rq_success_deferred_(false),
       upstream_wait_for_response_headers_before_disabling_read_(Runtime::runtimeFeatureEnabled(
-          "envoy.reloadable_features.upstream_wait_for_response_headers_before_disabling_read")) {
+          "envoy.reloadable_features.upstream_wait_for_response_headers_before_disabling_read")),
+      enable_tcp_tunneling_(enable_tcp_tunneling) {
```

Comment:
I think part of my confusion here is about the naming "enable tcp tunneling" when I feel like all the code should support tcp tunneling working all the time, so what are we enabling?
digging through code it looks to me like the only functional change across all this plumbing is to flush and force-(half) close with CloseForConnectRequestOrTcpTunneling on end stream.  so really this whole extra set of APIs is to avoid delay close when sending end stream FIN

I think largely we could kill this all off in favor of changing the HTTP/1.1 codec from endEncode to
 if (connect_request_ || is_tcp_tunneling_) {
to  if (connect_request_ || connection_.connection()).isHalfCloseEnabled() {

I think I'd prefer that change to plumbing more half close enabled all over.  it'd have to be its own runtime guarded change.  would you be OK with a fast turnaround PR doing that?  If not I can throw something together this week.

Reply:
sure, will send a PR for half-close-enabled first

Reply:
https://github.com/envoyproxy/envoy/pull/27834